### PR TITLE
Added default initialization to dash::Shared (with unit tests)

### PR DESCRIPTION
**Pull request by rkowalewski**

### CODE REVIEW

**Comment by devreal** [comment]
Thanks for the test case, I'll look at it later today.

**Comment by ddiefenthaler** [comment]
How are atomics initalized? As the values are signed, zero is not the absolute minimum.
If the start value is negative the minimum will be not the expected value.
From looking at the failed CI runs all compared values are negative.

I can't reproduce this locally. I guess in my machine the atomics are allocated in a zeroed memory segment. I think that is why I can't reproduce it.

**Comment by rkowalewski** [comment]
@ddiefenthaler you can either SSH into the CI machine or, like in my case, if you repeatedly execute it you will hopefully come into a situation which manifests the error.

**Comment by ddiefenthaler** [comment]
I added additional expects, which would be required to perform the rest of the test without problems.
The results support my assumption. The wrong value is already stored in the atomic _before_ the `min`&#8209;operation is called on it. I can't see an error in the rest of the unit test.

For the related issue in PR #543 I think there is missing a `barrier` which leads to fetching the minimum or maximum before it is fully computed.

Maybe there is still another problem, but I can neither see it in CI nor on my local machine.

**Comment by rkowalewski** [comment]
Ah now I got what you mean. Sorry, I didn't really know what I am doing here. Of course, we do not initialize the atomic value here, so it is undefined behavior. Stupid me. Actually, this refers to the same issue as in #479. This is something I am gonna change anyway in #515 , default initialization should be done for all containers. However, we can provide a temporary fix here, I am on the way to do this.

File: dash/include/dash/Shared.h [hunk 1]
```diff
@@ -83,7 +105,7 @@ class Shared {
       _owner,
       _team->dart_id());
     team.barrier();
```
Comment (devreal) [inline dash/include/dash/Shared.h]: That barrier is superfluous.

File: dash/include/dash/Shared.h [hunk 2]
```diff
@@ -58,6 +58,23 @@ class Shared {
...
+                   "finished delegating constructor");
+  }
+
+  /**
+   * Constructor, allocates shared value at single unit in specified team.
+   * The element is initialized with the given value.
+   */
+  Shared(
+    /// The value to initialize the element with
+    value_type    val,
```
Comment (devreal) [inline dash/include/dash/Shared.h]: Why not pass `val` by reference, i.e., `const valye_type&`?

**Review by devreal — CHANGES_REQUESTED** [review]

File: dash/include/dash/Shared.h (hunk 2 above)
Comment (ddiefenthaler) [reply dash/include/dash/Shared.h]: Good point, copied it over from the `set` function,
changed it there as well.

File: dash/include/dash/Shared.h (hunk 1 above)
Comment (ddiefenthaler) [reply dash/include/dash/Shared.h]: Thanks for the information.

I left it alone as it was there before I tempered with the file.
It is removed now.

**Review by devreal — APPROVED** [review]

**Comment by rkowalewski** [comment]
Thank you for your changes @ddiefenthaler.

**Pull request merged**